rasterio/_vsiopener.pyx

```diff
@@ -95,20 +95,28 @@ cdef void* pyopener_open(void *pUserData, const char *pszFilename, const char *p
         log.info("Object not found: registry=%r, filename=%r", registry, filename)
         return NULL
 
+    mode = pszAccess.decode("utf-8")
     cdef object file_obj
 
     try:
-        file_obj = file_opener(filename, mode="rb")
-    except ValueError:
-        file_obj = file_opener(filename, mode="r")
+        file_obj = file_opener(filename, mode)
+    # ZipFile.open doesn't take a mode argument and will raise
+    # ValueError if given one.
+    except ValueError as err:
+        file_obj = file_opener(filename)
 
-    # If the opener returned an OpenFile, we open it.
-    if not hasattr(file_obj, "f") and hasattr(file_obj, "fs") and hasattr(file_obj, "path"):
-        log.debug("Detected an OpenFile: file_obj=%r", file_obj)
-        file_obj = file_obj.open()
+    log.debug("Opened file object: file_obj=%r, mode=%r", file_obj, mode)
 
-    log.debug("Opened file object: file_obj=%r", file_obj)
-    _OPEN_FILE_OBJS.add(file_obj)
+    # Before we return, we attempt to enter the file object's context
+    # and store an exit callback stack for it.
+    stack = contextlib.ExitStack()
+
+    try:
+        file_obj = stack.enter_context(file_obj)
```

Comment:
@groutr @snowman2 this covers the fsspec OpenFile case and all other kinds of file objects.

Reply:
LGTM :)